Added shadow functionality for point lights.

mkkellogg opened the pull request:
Added shadow functionality for point lights.

mkkellogg committed:
Added shadow functionality for point lights.

WestLangley commented on examples/webgl_shadowmap_omnidirectional.html:
```diff
@@ -0,0 +1,295 @@
+				wallMaterial = new THREE.MeshPhongMaterial( {
+					color: 0xa0adaf,
+					shininess: 150,
+					specular: 0xffffff,
```
You may prefer shininess of 10 and specular of 0x111111, instead.

mkkellogg replied:
You're right, it looks better without the excessive shininess.

mrdoob commented:
Woh! This is great!

@bhouston @tschw do you guys mind double checking this a bit?

WestLangley commented on examples/webgl_shadowmap_omnidirectional.html:
```diff
@@ -0,0 +1,295 @@
+				scene.add( wall );
+				wall.position.set(-14,10,0);
+				rotateAroundWorldAxis(wall, new THREE.Vector3(0,0,-1), Math.PI / 180 * 90);
+
```
`wall.rotation.z = Math.PI / 2;` and remove `rotateAroundWorldAxis`

WestLangley commented on examples/webgl_shadowmap_omnidirectional.html:
```diff
@@ -0,0 +1,295 @@
+			    object.matrix = rotWorldMatrix;
+			    object.rotation.setFromRotationMatrix(object.matrix);
+			}
+
```
You have defined this twice. You do not need it, actually.

mkkellogg replied:
Ah, good catch, I'm not sure how that happened.

WestLangley commented:
Nice. + 1

I moved the point light inside the bounding box of the torus knot as an experiment. It looks good.

[mrdoobapproves](http://zz85.github.io/mrdoobapproves/) can help with your formatting if you are interested.

WestLangley commented on examples/webgl_shadowmap_omnidirectional.html:
```diff
@@ -0,0 +1,295 @@
+
+			function initScene() {
+
+				camera = new THREE.PerspectiveCamera( 45, window.innerWidth / window.innerHeight, 0.01, 1000 );
```
Small values for the near plane can be problematic. Try 1, instead.

mkkellogg replied:
Yeah I'd definitely be interested to hear about your formatting preferences.

WestLangley replied:
See the Wiki article [Mr.doob's-Code-Style™](https://github.com/mrdoob/three.js/wiki/Mr.doob%27s-Code-Style™).

mkkellogg replied:
Ah, thanks! I can definitely go back through my updates to make sure they conform style-wise.

mkkellogg replied:
The value for the 'near' plane on the point light's shadow camera has been updated to 1.

mkkellogg committed:
Updated code formatting.

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl:
```diff
@@ -1,12 +1,17 @@
 
+	uniform vec3 shadowLightPosition[ MAX_SHADOWS ];
+
 	varying vec4 vShadowCoord[ MAX_SHADOWS ];
```
No good to waste a uniform for a bool given what is discussed in #7037. 
Also, what is the difference between the `shadowLightPosition` and the position of the light passed in for lighting?

bhouston replied:
It also would be amazing to merge these into a vec4 someday:

```
 	uniform vec2 shadowMapSize[ MAX_SHADOWS ];
  	uniform float shadowDarkness[ MAX_SHADOWS ];
 	uniform float shadowBias[ MAX_SHADOWS ];
```

mkkellogg replied:
In this case, shadowLightPosition contains the world-space position of the light whereas pointLightPosition seems to contain the camera/view space position. I am currently doing all calculations for the point light shadows in world space, but I don't see why I couldn't update it to do them in camera/view space.

tschw replied:
@bhouston Yes, kind-of #7037...

@mkkellogg That sounds great! I moved the light positions to view space just very recently (see #7028) and there were no references in the shadow mapping code... Is it possible to get away without altogether perhaps? Either way, no problem when reusing the lighting info...

mkkellogg replied:
So as it turns out, all the calculations can still be done in world space (which makes things much easier) as long as the point light's position is converted to world space in the vertex shader. Since I was previously not making use of the 'shadowMatrix' uniform for point lights, I can use that matrix to perform the conversion by populating it with the inverse of the view matrix. Does that sound reasonable?

tschw replied:
> So as it turns out, all the calculations can still be done in world space (which makes things much easier) as long as the point light's position is converted to world space in the vertex shader. Since I was previously not making use of the 'shadowMatrix' uniform for point lights, I can use that matrix to perform the conversion by populating it with the inverse of the view matrix. Does that sound reasonable?

Note sure: It's a high base cost for the matrix. Without it, we get 4 light positions for "free", right?


Also beware the number of varying vectors. See `MAX_VARYING_VECTORS` and `MAX_FRAGMENT_UNIFORM_VECTORS` on [webglstats.com](http://webglstats.com/), switching off anything but "Smartphone".

> As for the 'isShadowCube' uniform, into which other existing uniform would you prefer that be packed?

I would probably end up just using the sign of an unsigned value and add a comment...

mkkellogg replied:
Ah, so are you thinking of doing away with 'shadowMatrix'? I thought it would be OK to use it in the alternate manner that I described above since it wouldn't add any extra computations over the existing shadow mapping code (in the vertex shader). Basically the vertex shader would look something like:

```
//  shadowDarkness[ i ] < 0 means we have a point light
if( shadowDarkness[ i ] < 0 ){

     vShadowCoord[ i ] = shadowMatrix[ i ] * vec4( pointLightPosition[ i ], 1.0 );

} else {

     vShadowCoord[ i ] = shadowMatrix[ i ] * worldPosition;

}
```

Additionally, I would be re-purposing 'vShadowCoord' to hold the world-space position of the light. I admit it's not the most readable code, but I'd make sure to comment appropriately.

tschw replied:
> Ah, so are you thinking of doing away with 'shadowMatrix'?

I think we should avoid requiring higher limits. Especially in case the new feature is not being used. There is always the option of using a define to ensure the latter. It may be a good idea in any case to get rid of the cube sampler (*1) when it's not needed, however, when that's the only purpose I'd say we have a smooth integration.

The story of GL ES can be summarized as follows: `MAX_VARYING_VECTORS` can be as low as 8 and `MAX_FRAGMENT_UNIFORM_VECTORS` can be as low as 16. Only *statically referenced* variables count, meaning if you don't touch the variable they won't be allocated (the cleverness to detect dead code may vary across different drivers). In theory, there should also be some automatic packing, which we can't rely on in practice. 

> Without it, we get 4 light positions for "free", right?

No. I was wrong here. Using a matrix in the vertex shader is not much of a problem, however,

>Additionally, I would be re-purposing 'vShadowCoord' to hold the world-space position of the light. I admit it's not the most readable code, but I'd make sure to comment appropriately.

adding extra varyings is. => Using `vShadowCoord` seems a very good idea.

Comments are nice, and even nicer is to "say it in code" when possible (comments easily get outdated by edits, code OTOH has to be dealt with), e.g.:

```javascript
bool isPointLight = shadowDarkness[ i ] < 0;
// ...
```

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_fragment.glsl:
```diff
@@ -26,148 +31,165 @@
+				if( currentIsCube == 1 ){
+				
+					shadow = sampleCubeShadowMapPCF( shadowCube[ i ], lightToFragment, distanceToLight, cubeTexelSize, 1.5);
+					
```
Some drives may not like passing a shader to a function.

tschw commented:
Nice example! :+1: 

With an Intel Iris GPU using the Intel-contributed driver in Mesa under Linux I see:

THREE.WebGLShader: gl.getShaderInfoLog()" "fragment" "0:182(55): warning: sampler arrays indexed with non-constant expressions will be forbidden in GLSL 1.30 and later
0:188(41): warning: sampler arrays indexed with non-constant expressions will be forbidden in GLSL 1.30 and later

I guess it will disappear when you get rid of passing the sampler into a function. Unfortunately the line numbers are off, but that appears an unrelated issue to me (I suspect either the driver or Three.js dev head).

mkkellogg committed:
Updated point-light shadow mapping code to be more efficient with uniforms and varyings.

mkkellogg committed:
Inlined code from sampleCubeShadowMapPCF() directly into main().

mrdoob commented:
I'll wait or @tschw approval on this one 😊

tschw commented:
Almost :-)

All code should be wrapped with something like `#if defined(POINT_LIGHT_SHADOWS)`, where this macro is only defined when there are actually point lights that cast shadows in the scene.

We have to do it, because the presence of the new feature will currently allocate twice the number of samplers (than actually used). There is no way for the compiler to determine which array elements are being referenced by static analysis, and there is no sufficiently simple integration of the provided code that overcomes this problem.

Once that's done, I think it's reasonable to merge this one.

To achieve optimal resource usage we may want to consider a software projection to get rid of the cube sampler.

WestLangley commented on src/renderers/shaders/UniformsLib.js:
```diff
@@ -117,13 +117,15 @@ THREE.UniformsLib = {
 
-		"shadowMatrix" : { type: "m4v", value: [] }
+		"shadowMatrix" : { type: "m4v", value: [] },
 
```
comma

mkkellogg replied:
I will remove this.

WestLangley commented on src/renderers/webgl/WebGLShadowMap.js:
```diff
@@ -153,83 +224,105 @@ THREE.WebGLShadowMap = function ( _renderer, _lights, _objects ) {
-			//
+					_lookTarget.copy( shadowCamera.position );
+					_lookTarget.add( cubeDirections[face] );
+					shadowCamera.up.copy( cubeUps[face] );
```
formatting

mkkellogg replied:
I'm not sure what you mean about this one. It seems like this file is formatted appropriately. I double checked using this tool: http://zz85.github.io/mrdoobapproves/

WestLangley commented:
This is nice.

Any idea why `ShaderLib.js` has been completely replaced?

This will conflict with prior PRs, so you will have to resolve that.

mkkellogg committed:
Added checks for definition of POINT_LIGHT_SHADOWS in shaders for including point light shadow mapping code.

mkkellogg commented:
I actually have no idea why ShaderLib.js completely changed, but I went ahead took the upstream version and re-added my changes to that file.

WestLangley commented:
I previously made three comments that you have so far failed to respond to. They are easy fixes.

You should not be including the builds in the PR. See [How to Contribute to three.js](https://github.com/mrdoob/three.js/wiki/How-to-contribute-to-three.js).

Thanks.

mkkellogg commented:
@WestLangley - Sorry about missing those comments, although I think I only see two that I missed. Is the third comment the one you initially made regarding the replacement of ```rotateAroundWorldAxis``` with ```wall.rotation.z = Math.PI / 2;``` in examples/webgl_shadowmap_omnidirectional.html? If so, I have taken care of that update.

tschw commented:
@mkkellogg 

I implemented cube mapping in a few lines of (branchless) GLSL code to get rid of the cube sampler array. With this [input image](http://www.kweepa.com/step/ags/tech/Heartland360.jpg) my testbed produces
![render2](https://cloud.githubusercontent.com/assets/3916357/9941300/a66d49b2-5d74-11e5-9782-c983913e0f97.jpg)

or, using an [annotated input image](http://learnopengl.com/img/advanced/cubemaps_skybox.png) (note the seams in the image - these are not rendering artifacts)
![render1](https://cloud.githubusercontent.com/assets/3916357/9941298/a2390bec-5d74-11e5-8f6e-ec2ca9778421.jpg). The output is from a single `texture2D` invocation after a coordinate conversion from a `textureCube`-style coordinate vector. The images were produced with bilinear filtering enabled (the routine can optionally avoid reading from the danger zone).

**In case you have to tidy up the changes on your `dev` branch**
```shell
$ git branch UpstreamDev
$ git checkout UpstreamDev
$ git reset --hard <random commit before your changes came in>
$ git pull upstream dev
$ # UpstreamDev is now in sync with upstream dev
$ git checkout dev 
$ # back with the changes
$ git rebase --interactive UpstreamDev
```
and select the commits to pull *on top* of `dev`. You may want to edit some commits that added stuff you don't want to submit.

**Note that** (don't enter these lines as a script, they're for your information and to make sense of what git may tell you in the process)
```shell
$ git reset HEAD # unstages all /add/ed changes
$ git commit --amend # overwrites the previous commit with a new one
$ git reset <earlier> # undoes commits but leaves your working copy unchanged
$ git reset --hard <earlier> # also resets your working copy
```

**Next time**, you may have an easier life avoiding to use your `dev` branch for the PR, IOW:
```shell
$ git branch MyNewFeature # create feature branch
$ git checkout MyNewFeature # switch to feature branch
```
. Then you can always rewrite your entire history saying
```shell
$ git reset dev
$ # ... /add/ and /commit/ your changes again
```
(it requires your changes to be on top of the upstream ones).

At some point you may want to sync with upstream. In this case:
```shell
$ git checkout dev # switch back to older point in upstream history
$ git pull upstream dev # do a painless "fast-forward" update
$ git checkout MyNewFeature # back to feature branch
$ git rebase dev # pull your changes on top of updated upstream
```

Note that you can easily rewrite your entire history again. Also note that you're best off to *never* "*pull* upstream dev" into your feature branch.

tschw commented:
I put the cubemapping code [here](https://gist.github.com/tschw/da10c43c467ce8afd0c4) in case you want to use it. More sophisticated options are [dual parabolic](http://osman.brian.googlepages.com/dpsm.pdf) and tetrahedral projections. The former are very fast to render (two shots) but yield artifacts with too little tessellation. The latter do not suffer this limitation, but are not quite as fast (four shots) and will need custom frustum culling to avoid redundant rendering.

mkkellogg commented:
@tschw Thanks! This will be quite helpful in my implementation.

tschw commented:
I used the horizontal cross layout because there are existing images. The mechanisms in the code work for pretty much any layout. For what we're up to, we could go up and down on different squares:
```
Y            Uppercase: Positive side
xzXZ         Lowercase: Negative side
  y          Character: Axis
```
With REPEAT wrapping, we can rearrange it to
```
Y y
xzXz
```
and get square texels on a 2:1 texture (no anisotropic distortion). Also, we'd use 75% of the image and not just half of it.

mkkellogg commented:
@tschw I have implemented the shadow mapping with the cube-to-2D projection using a 2:1 texture with the final orientation you suggested:

```
Y y
xzXz
```

I created a new branch where you can see the updates [here](https://github.com/mkkellogg/three.js/tree/feature/omnidirectional_shadowmaps). I updated the cubeToUV() function to account for the new 2:1 arrangement, but it didn't seem to me that REPEAT wrapping was necessary (that probably means I did something wrong). Please let me know what you think. 

Thanks.

mrdoob commented:
> if you prefer I can just put the changes into `dev` and update the pull request

I think that would be better, so we can easily track the changes in this same PR.

tschw commented:
@mkkellogg Oh, you implemented it already! Very nice!

I ended up implementing
```
xzXZ
 y Y
```
because I found it easiest to validate: To construct a texture image from a horizontal cross simply move the top square.

> it didn't seem to me that REPEAT wrapping was necessary (that probably means I did something wrong)

No. The REPEAT wrapping idea didn't work out too well for me either :-). But the skipping of the seams was somewhat hacky and your code inherited the hackiness.

The new code can be found in the same Gist ( https://gist.github.com/tschw/da10c43c467ce8afd0c4 ) in the file `cubeToUVCompact.glsl`. Please put the link into a comment. This way, whoever wants to edit it can find a standalone test case.
I also updated the horizontal cross variant with proper seam skipping, scaling all values in respect to the center of each square and considering anisotropy. The compact variant does so too, but there is no anisotropy, so in that case it's a bit simpler. The new code is no longer branchless, since many GPUs implement flow control and the divergent case (when at the edges) is relatively rare.

BTW check your editor / local git configuration - your commits get diffed much more than they should. Maybe about line endings...

mkkellogg commented:
@tschw I have committed my latest changes to dev. I apologize for all the merge commits cluttering up the history, I'll try to rebase my updates from here on out. 

As far as the excessive diffs are concerned, I think you're right about the line endings. I have core.autocrlf=true since I'm working in windows, but git should convert to LF upon commit. I think I will change that to false and be extra diligent about preventing it in the future.

WestLangley commented:
@mkkellogg The comments I made that you seem to have missed are (1) to avoid small values of the `near` plane (I suggested a value of 1), and (2) not to include the builds in the PR.

tschw commented:
@mkkellogg 
Recipe: Rebase, reset before your commits, ensure consistent line endings on all edited files, add & commit again, push --force origin dev.

tschw commented:
> I apologize for all the merge commits cluttering up the history, 

No need to apologize. Cluttering the history is a lesser concern than burying changes underneath that you may want to uncommit. I was not trying to complain, but to be helpful. However, I'm used to that common misunderstanding :-).

mrdoob commented:
Do you mind excluding the builds (three.js and three.min.js) from the PR? Mainly for clarity for also to avoid collisions.

mkkellogg committed:
Point light shadow mapping now utilizes cube-to-2D projection, avoiding the need for a samplerCube in the shader.

mkkellogg commented:
@mrdoob @tschw  I rebased my ```dev``` branch and made sure that all modified files use LF line endings. I also made sure to exclude the builds this time. 

@WestLangley I have gone over all the small changes you have suggested and updated ```dev``` to reflect them, so everything should be in order.

mkkellogg committed:
Updated comments for cubeToUV() function in shadow mapping shader.

tschw commented:
@mkkellogg 

Reads much better! As @WestLangley suggested, `near` should be set to 1 - it's more robust with lower precision hardware.

WestLangley commented on src/renderers/shaders/ShaderChunk/shadowmap_vertex.glsl:
```diff
@@ -2,7 +2,36 @@
+				// When we have a point light, the @shadowMatrix uniform is used to store
+				// the inverse of the view matrix, so that we can get the world-space
+				// position of the light.
+
```
I am not a bit fan of repurposing, in this case, and elsewhere in this PR, but of course, that is a matter of opinion.

In this case, the repurposing is not necessary, as we already have access to the `viewMatrix`. We have used a similar trick previously in `common.glsl` to transform direction vectors by the inverse of the view matrix.

See https://en.wikibooks.org/wiki/GLSL_Programming/Applying_Matrix_Transformations, and in particular the section _Transforming Points with the Inverse Matrix_. 

@tschw Can you confirm?

tschw replied:
Yes. Sorry for not catching that one. In fact, you can find the same trick in the test for the software cubemapper transforming the ray. Inverse, of course...

Since we assume view <-> world to be orthonormal, we can even go without the normalization - just multiply from the other side.

mkkellogg replied:
Ah, I didn't know there was necessarily an orthonormal guarantee for the view matrix. I originally included the inverse of the view matrix in ```shadowMatrix ``` because I didn't want to calculate the inverse in the vertex shader. But as @tschw mentioned, since the view matrix is orthonormal we can just use the transpose as the inverse, or even better use pre-multiplication instead of post :)

@tschw You mentioned that I can use the same trick in the software cubemapper... are you saying I should avoid re-purposing vShadowCoord altogether and calculate the ray in the fragment shader? This would mean a lot of extra matrix multiplications..is that acceptable?

tschw commented:
@mkkellogg No. I just picked up the term "trick" from @WestLangley referring to the routine in `common.glsl` for inverse multiplication, which reads somewhat like this (cited from memory):
```glsl
return normalize( vec4( direction, 0.0 ) * matrix );
```
By now, we can simply use `direction * mat3( matrix )`. 

Mentioning the test code was probably misleading: As we all know, rays are inverse-transformed compared to the matrices used for rasterization, and so are camera transforms. However, since I called the variable `camAngles`, one could even say I missed the second inversion cancelling out the first - but it doesn't really matter in that case :-).

@WestLangley 
> I am not a bit fan of repurposing, in this case, and elsewhere in this PR

I think neither of us really is, but statically referenced uniforms are not ethereal variables - they affect the scalability and portability of the library. If it's about something else, please be more specific. If you want to disagree with me, let's keep other contributors out of it and discuss it (ideally before we have an implementation on the table, but that's too late in this particular case, or) for a cleanup pass after the merge :-).

mkkellogg committed:
Updated point light shadow map shader to use existing 'viewMatrix' and 'shadowBias' variables.

mkkellogg commented:
@tschw The code in ```shadowmap_vertex.glsl ``` now uses ```viewMatrix``` to transform the distance vector into world space.

WestLangley commented:
@tschw 

> I am not a big fan of repurposing, in this case, and elsewhere in this PR
>> I think neither of us really is, but statically referenced uniforms are not ethereal variables 

Yes, I was referring to the repurposing of uniforms. I will defer to your decision on this one.

> As @WestLangley suggested, near should be set to 1 - it's more robust with lower precision hardware.

Yes, I indeed suggested that 12 days ago. I am having a hard time getting my comments to be recognized.

mkkellogg commented:
@WestLangley 

> > As @WestLangley suggested, near should be set to 1 - it's more robust with lower precision hardware.

> Yes, I indeed suggested that 12 days ago. I am having a hard time getting my comments to be recognized.

I'm sorry for any confusion, I made this update immediately after you suggested it 12 days ago. I will make sure to respond directly to every comment going forward.

WestLangley commented:
@mkkellogg Sorry, I now see we have been cross-talking, and in your view, you responded to my concerns.

My original comment was in regards to this line: I would set the near plane to 1.
```javascript
camera = new THREE.PerspectiveCamera( 45, window.innerWidth / window.innerHeight, 0.01, 1000 );
```
You also have the following line commented out. It is OK, however, because the shadow camera is orthographic. But still, users may not understand the subtlety, so I would change it to 1, if you agree.
```javascript
dirLight.shadowCameraNear = 0.01;
```
The line below, for `PointLight`, I expect is also problematic. But to be honest, I haven't had the time to study it in detail. ( /ping @tschw )
```javascript
this.shadowCameraNear = 0;
```


I must say, IMO you have done an excellent job here. I also appreciate your descriptive inline comments. : - )

mkkellogg commented:
@WestLangley 

> My original comment was in regards to this line: I would set the near plane to 1.

> ```camera = new THREE.PerspectiveCamera( 45, window.innerWidth / window.innerHeight, 0.01, 1000 );```

Ah ok, that makes sense. I think I inherited this line when I copied the existing shadow mapping example and just subconsciously overlooked it ( that's the excuse I'll go with anyways :) ). I will update it (I can update the example from which it was copied as well if you want).

> You also have the following line commented out. It is OK, however, because the shadow camera is orthographic. But still, users may not understand the subtlety, so I would change it to 1, if you agree.

> ```dirLight.shadowCameraNear = 0.01;```

I agree, even though it is commented out, it should be updated to 1 for consistency.

>The line below, for PointLight, I expect is also problematic. But to be honest, I haven't had the time to study it in detail. ( /ping @tschw )

> ```this.shadowCameraNear = 0;```

I think you're right about this one as well. The assumption here was that the developer would always appropriately set the values before usage, but the default values should be more reasonable. I can change this to 1 as well.

> I must say, IMO you have done an excellent job here. 

Thank you! :)

@WestLangley @tschw These updates will be quick and easy, is there anything else I should address in my next update to the PR?

mrdoob commented:
> I must say, IMO you have done an excellent job here. I also appreciate your descriptive inline comments. : - )

Ditto!

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl:
```diff
@@ -16,4 +16,136 @@
+			vec2 planar = v.xy;
+
+			// stop at the last texel (can use a factor of 1.0 for NEAREST)
+			float yTexelSize = 2.0 * Squares.y * texelSizeY;
```
Filtering is Nearest. So shouldn't have to swallow pixels here...

mkkellogg replied:
Ah, right, I will fix that.

mkkellogg replied:
@tschw I updated the shadow map rendering code to use the same compact layout as you did in your cubeToUVCompact() function:

```
// xzXZ
//  y Y
```

I like this layout better since it requires only one move to get to a horizontal cross layout. I updated the existing branch-less cubeToUV() function to use this layout, and to use a better approach to seam skipping, and you can see that code [here](https://gist.github.com/mkkellogg/02db99420dfd105513e9). I also updated the shadow map rendering code to use linear filtering and it still seems like I can use a factor of 1.0 here:

```
	// stop at the last texel (can use a factor of 1.0 for NEAREST).
	float yTexelSize = 1.0 * Squares.y * texelSizeY;
```

I tried several shadow map resolutions, down to 128 x 64 and it didn't look like there were any seam-related issues.

Your cubeToUVCompact() function is still much easier to read than the branch-less version - would you rather that I use it over the branch-less code?

tschw replied:
> and you can see that code here

It reads pretty similar to my version before I reimplemented the priority decoder via `if` .. `else if` - and similarly obscure.

> Your cubeToUVCompact() function is still much easier to read than the branch-less version - would you rather that I use it over the branch-less code?

Yes, readability matters - but it's not just about that:

The reason I changed it was because I realized that there would be no benefit from branchless code and it had become an end in itself: 

Calculating lots of stuff that is cancelled out plus doing additional stuff to avoid branches is actually worse than what we'd get from compiling a few `if`s for a target that can't jump. The latter implementation takes fewer instructions in total!

Furthermore, many GPUs *can* jump and use an [SIMT execution model](https://en.wikipedia.org/wiki/Single_instruction,_multiple_threads), IOW they really skip code that is not needed by any of the (typically 32) lock-stepping threads in the same execution vector processing nearby fragments. Hardware of this type benefits most from the branching variant.

I don't use a separate zoom factor for x, just assuming 2:1. In this case, we don't need it, since `texelSizeX` is half of `texelSizeY` (UV space) but we have twice the number of squares along the axis => same result.

> I also updated the shadow map rendering code to use linear filtering and it still seems like I can use a factor of 1.0

Tested & confirmed. I lowered that factor in the [code](https://gist.github.com/tschw/da10c43c467ce8afd0c4).

mkkellogg replied:
> Calculating lots of stuff that is cancelled out plus doing additional stuff to avoid branches is actually worse than what we'd get from compiling a few ifs for a target that can't jump. The latter implementation takes fewer instructions in total!

> Furthermore, many GPUs can jump and use an SIMT execution model, IOW they really skip code that is not needed by any of the (typically 32) lock-stepping threads in the same execution vector processing nearby fragments. Hardware of this type benefits most from the branching variant.

Given this, I think we should use your updated code. I replaced the existing code with your implementation of cubeToUVCompact() and it works fine. I will include that in the next update to the PR.

Are there any other issues I should address before updating the PR?

tschw replied:
Not that I am aware of.

Can't get enough of it? No problem - doesn't have to be your last PR. Let's get this one merged :-).

mkkellogg replied:
Alright, the latest changes are committed, so it should (hopefully) be ready to go.

WestLangley commented:
> I can update the example from which it was copied as well if you want

Please do.

mkkellogg committed:
Updated cubeToUV() to a reduced-instruction version with minimal branching, and updated near plane values for shadow map cameras to be at least 1.

mkkellogg commented:
@tschw @WestLangley 

The PR now has the latest updates, including the update to cubeToUV() to use branching and updates to near plane values that were too small. It should (hopefully) be ready to go now.

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl:
```diff
@@ -16,4 +16,124 @@
+		*  Z - Positive z direction
+		*  z - Negative z direction
+		*
+		*  Alternate code for a horizontal cross layout can be found here:
```
I'd prefer it to read: "Original source **and testbed**:"

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl:
```diff
@@ -16,4 +16,124 @@
+		*
+		*  Alternate code for a horizontal cross layout can be found here:
+		*  https://gist.github.com/tschw/da10c43c467ce8afd0c4
+		*/
```
Also, more generally, though not a strong requirement: When having the choice, it's a good idea to opt for single-line comments, so the multi-line ones can be used to disable code. Less of an issue in GLSL (where there is `#if 0` ... `#endif`) than in JavaScript below.

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl:
```diff
@@ -16,4 +16,124 @@
+		bool gridSamplingInitialized = false;
+
+		void initGridSamplingDisk(){
+
```
Space.

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl:
```diff
@@ -16,4 +16,124 @@
+		void initGridSamplingDisk(){
+
+			if( gridSamplingInitialized ){
+
```
Space.

tschw commented on src/renderers/webgl/WebGLShadowMap.js:
```diff
@@ -98,13 +152,58 @@ THREE.WebGLShadowMap = function ( _renderer, _lights, _objects ) {
 				var shadowFilter = THREE.LinearFilter;
 
-				if ( scope.type === THREE.PCFSoftShadowMap ) {
+				if ( scope.type === THREE.PCFSoftShadowMap) {
```
Chewed-off space.

tschw commented on src/renderers/webgl/WebGLShadowMap.js:
```diff
@@ -55,6 +70,43 @@ THREE.WebGLShadowMap = function ( _renderer, _lights, _objects ) {
+	_distanceMaterialMorph._shadowPass = true;
+	_distanceMaterialSkin._shadowPass = true;
+	_distanceMaterialMorphSkin._shadowPass = true;
+
```
This routine could be a lot DRYer:

```javascript
var MorphTargets = 1, Skinning = 2;
var nMaterials = MorphTargets | Skinning;

var _depthMaterials = new Array( nMaterials );
var _distanceMaterials = new Array( nMaterials );

for ( var i = 0; i !== nMaterials; ++ i ) {

    var useMorphTargets = ( i & MorphTargets ) !== 0;
    var useSkinning = ( i & MorphTargets ) !== 0;

    _depthMaterials[ i ] = new THREE.ShaderMaterial( {
        uniforms: depthUniforms,
        vertexShader: depthShader.vertexShader,
        fragmentShader: depthShader.fragmentShader,
        morphTargets: useMorphTargets,
        skinning: useSkinning
    } );

    _distanceMaterials[ i ] = new THREE.ShaderMaterial( {
        uniforms: distanceUniforms,
        vertexShader: distanceShader.vertexShader,
        fragmentShader: distanceShader.fragmentShader,
        morphTargets: useMorphTargets,
        skinning: useSkinning
    } );

}
```

Not sure @mrdoob will like it, though....

tschw commented on src/renderers/webgl/WebGLShadowMap.js:
```diff
@@ -246,44 +365,78 @@ THREE.WebGLShadowMap = function ( _renderer, _lights, _objects ) {
 		var useMorphing = geometry.morphTargets !== undefined && geometry.morphTargets.length > 0 && material.morphTargets;
 		var useSkinning = object instanceof THREE.SkinnedMesh && material.skinning;
 
-		var depthMaterial;
```
DRYer variant:

```javascript
// [... handle object.custom*, see below ... ]

var materialVariants ? isPointLight ? _distanceMaterials : _depthMaterials;
vat index = ( useMorphing ? MorphTargets : 0 ) | ( useSkinning ? Skinning : 0 );
return materialVariants [ useMorphing ? MorphTargets : 0 )
```

mrdoob commented:
Merging! We can do cosmetic changes in a new PR.

mrdoob commented:
Many many thanks guys!

tschw commented on src/renderers/webgl/WebGLShadowMap.js:
```diff
@@ -246,44 +365,78 @@ THREE.WebGLShadowMap = function ( _renderer, _lights, _objects ) {
+
+				newMaterial = object.customDepthMaterial;
+
+			}
```
Looking at a subtle issue, here: The presence of either one of the properties may cause the other one, possibly `undefined`, to be assigned to `newMaterial`.

```javascript
var customMaterial = isPointLight ?
        object.customDistanceMaterial :
        object.customDepthMaterial;

if ( customMaterial !== undefined ) {

    newMaterial = customMaterial;

}
```

tschw replied:
I addressed that one in #7236.

tschw commented on src/renderers/webgl/WebGLShadowMap.js:
```diff
@@ -246,44 +365,78 @@ THREE.WebGLShadowMap = function ( _renderer, _lights, _objects ) {
+
+			}
+
+		}
```
I guess it makes sense to require this uniform on custom depth materials because they will need it?

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_fragment.glsl:
```diff
@@ -26,148 +41,252 @@
+								// nested loops breaks shader compiler / validator on some ATI cards when using OpenGL
+								// must enroll loop manually
+							for ( float y = -1.25; y <= 1.25; y += 1.25 )
+								for ( float x = -1.25; x <= 1.25; x += 1.25 ) {
```
Anyone remember details / when it was / has an affected GPU for testing? /ping @WestLangley @mrdoob 
I guess the issue in this comment *may* be specific to the index being of type `float`?

mrdoob replied:
We could try removing it and see if someone complains.

tschw replied:
We didn't consider the bug for point shadows, so these may or may not work with that GPU. 

We're safe when the new code is not being used - in that case, the macro will keep it from being compiled in, so existing code should keep working.

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl:
```diff
@@ -16,4 +16,124 @@
+			// Apply scale to avoid seams
+
+			// two texels less per square (one texel will do for NEAREST)
+			v *= scaleToCube * ( 1.0 - 4.0 * texelSizeY );
```
@mkkellogg 

The new version (updated yesterday, before my post) reads:
```javascript
// two texels less per square (one texel will do for NEAREST)
vec2 zoom = 1.0 - 2.0 * vec2( 4.0, 3.0 ) * texelSize;
```
Did you intentionally use the older one because there were artifacts? Or did you just make the changes before the update?

mkkellogg replied:
I think this was simply an oversight on my part, I referenced the older code when I made the final update. There were no artifacts when using the new version.

mrdoob commented:
While I was improving the example, I noticed that there seem to be a bug when uncommenting the dirLight code.

https://github.com/mrdoob/three.js/blob/dev/examples/webgl_shadowmap_pointlight.html#L88-L103

![screen shot 2015-09-25 at 10 05 46](https://cloud.githubusercontent.com/assets/97088/10102588/41fb0a24-636d-11e5-94ba-ae496e7ca5c2.png)

![screen shot 2015-09-25 at 10 06 21](https://cloud.githubusercontent.com/assets/97088/10102605/6dcecb40-636d-11e5-8ccc-3f4adebadeca.png)

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl:
```diff
@@ -16,4 +16,124 @@
+			gridSamplingInitialized = true;
+			
+		}
+
```
@mkkellogg 

It takes a very advanced GLSL compiler to track these values as constants to the point where they are finally being used - a more advanced compiler than can be assumed, considering the intentional limitations in GLSL ES v1.

bhouston replied:
Does OpenGL ES 1.0 support "const array" declarations outside of any function?  If I remember correctly it doesn't, only individual floats or vectors are supported.

tschw replied:
Exactly. This is in fact one of the limitations I was referring to. I think it makes sense to manually unroll the loop that consumes these constants and inline them.

mkkellogg replied:
I was aware of this limitation of GLSL ES v1; I just wanted to avoid unrolling a 20 iteration loop. However, performance considerations should probably be a higher priority over a little extra code repetition.

Also, given the existence of the AMD bug, unrolling might be necessary.

tschw replied:
The point is, that when the compiler fails to see these array elements as constants, it needs to (statically) allocate, write and read per-thread memory. 

The static allocation can limit the number of parallel threads that the GPU can process at once. If it doesn't, we're talking about out-of-core memory and uncached access is slow. In case there's caching, the caches (also used for other things) are polluted with the same data over and over. Neither of these cases is pleasant.

> Also, given the existence of the AMD bug, unrolling might be necessary.

Oh well, imperfection and limitations are part of the deal when writing software. Not saying it's a crime to be ambitious :-).

Congratulations on the merge for now!

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_fragment.glsl:
```diff
@@ -26,148 +41,252 @@
+
+					if( isPointLight ) {
+
+						float cubeTexelSize = 1.0 / ( shadowMapSize[ i ].x * 0.25 );
```
same as `4.0 / shadowMapSize[ i ].x` == `4.0 * texelSizeX` == **`2.0 * texelSizeY`**

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_fragment.glsl:
```diff
@@ -26,148 +41,252 @@
+							numSamples += 1.0;
+
+						}
+
```
Considering the AMD bug (mentioned in the comment below) and the array of constants, I think it makes sense to just unroll this loop.

tschw commented on src/renderers/shaders/ShaderChunk/shadowmap_fragment.glsl:
```diff
@@ -26,148 +41,252 @@
+
+						initGridSamplingDisk();
+
+						float diskRadius = 1.25;
```
Actually
```glsl
const float DiskRadius = 1.25;
```

tschw commented:
@mkkellogg 

Would you mind having a look at that bug reported by @mrdoob above? 
Also there's #7239 that might be related...

mkkellogg commented:
@tschw Sure, I will look into it.

mkkellogg commented:
@tschw @mrdoob 

> While I was improving the example, I noticed that there seem to be a bug when uncommenting the dirLight code.

This is a very simple bug to fix. In WebGLShadowMap.js, this line:

``` _renderer.setViewport( _vector4.x, _vector4.y, _vector4.z, _vector4.w ); ```

needs to be moved to the end of this loop:

```for ( var i = 0, il = _lights.length; i < il; i ++ ) {```

Would you like me to make the change?

mrdoob commented:
> Would you like me to make the change?

Yes please 😃

mrdoob commented:
@mkkellogg @tschw 

Any idea why these glitches appear on iOS?

![image](https://cloud.githubusercontent.com/assets/97088/12691239/5141ecd6-c72d-11e5-827d-8d9f0e1728d8.png)

http://threejs.org/examples/#webgl_shadowmap_pointlight

This is with the r73 so it's not my modifications...

I'm suspicious of cubeToUV.

tschw commented:
@mrdoob 
> Any idea why these glitches appear on iOS?

Looks like the depth calculation running out of precision or depth value encoding quirks off.

> I'm suspicious of cubeToUV.

Here's my keen bet against your suspicion: cubeToUV is just a coordinate transform and does not mess with the values encoded at each texel at all (recall its use `texture2D( mySampler, cubeToUV( direction ) )` ), but the screenshot looks more like the texel positions are just fine but the values found at those texels glitch off.
 
My bet is on the [`distanceRGBA`](https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderLib.js#L1047-L1111) shader pair.

It seems the routine that does the encoding has changed in [`depthRGBA`](https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderLib.js#L981-L1044) - the corresponding code for light sources other than point lights. The line
```glsl
vec4 res = fract( depth * bit_shift );
```
has been commented out. Guessing it may have caused similar artefacts for other light sources. I'd try to copy the change over for point lights...

mrdoob commented:
Interesting... I'll take a look. Thanks!
Also, we have that bit_shift mask repeated all over the place. Time to do some clean up.

mrdoob commented:
> Guessing it may have caused similar artefacts for other light sources. I'd try to copy the change over for point lights...

That did it! Many thanks!

mrdoob commented:
Seems like I'm quickly hitting the limits of the current approach:

https://twitter.com/thespite/status/697013941386477569

As far as I understand after reading http://www.sunandblackcat.com/tipFullView.php?l=eng&topicid=36 most of the [current 21 texture fetches](https://github.com/mrdoob/three.js/blob/dev/src/renderers/shaders/ShaderChunk/shadowmap_pars_fragment.glsl#L219-L239) are redundant.

> To sample only required direction it's possible to calculate tangent vectors that will determine base axes for offset vectors. First we need to get a perpendicular to main sampling direction. Let's consider it as X axis for offset. It can be determined as cross product of main sampling direction and an arbitrary vector (not parallel to main sampling direction). Then we can calculate vector for Y offset axis that is perpendicular to main sampling direction and to X offset axis. These two perpendicular vectors can be used as base X and Y axes for offset. You can even precompute 2D kernel (grid, poisson, randomly rotated, etc.) with offsets as for two-dimensional shadow map.

Anyone up for implementing the proposed solution?

mrdoob commented:
Now using a 4 tap approach.

bhouston commented:
BTW Sketchfab uses both esm and vsm shadows:

http://http.developer.nvidia.com/GPUGems3/gpugems3_ch08.html

http://developer.download.nvidia.com/presentations/2008/GDC/GDC08_SoftShadowMapping.pdf

mkkellogg commented:
@mrdoob 

I looked over your 9-sample approach and it looks like you are not calculating the orthogonal vectors to be used as a new basis for the offset vectors (as suggested [here] ( http://www.sunandblackcat.com/tipFullView.php?l=eng&topicid=36)).  Was that intentional, or am I missing something? Using the approach described in that link I came up with this:

```javascript

// a is an arbitrary vector whose only purpose is to NOT be parallel with bd3D
vec3 a = bd3D + vec3( 1, 0, 0 ) + vec3( 0, 1, 0 );
// create new basis x & y vectors that are orthogonal to each other and to bd3D
vec3 ty = normalize( cross( a, bd3D ) );
vec3 tx = normalize( cross( bd3D, ty ) );

// the extra multiplication by 2.0 at the end is just for experimenting to get nice soft edges,
// ultimately that should be rolled into "shadowRadius" 
vec4 offset = vec4( - 1, - 0.707107, 1, 0.707107 ) * shadowRadius * 2.0 * texelSize.y * 2.0;

return (

	texture2DCompare( shadowMap, cubeToUV( bd3D, texelSize.y ), dp ) +
	texture2DCompare( shadowMap, cubeToUV( bd3D + tx * offset.w + ty * offset.w, texelSize.y ), dp ) +
	texture2DCompare( shadowMap, cubeToUV( bd3D + tx * offset.y + ty * offset.w, texelSize.y ), dp ) +
	texture2DCompare( shadowMap, cubeToUV( bd3D + tx * offset.y + ty * offset.y, texelSize.y ), dp ) +
	texture2DCompare( shadowMap, cubeToUV( bd3D + tx * offset.w + ty * offset.y, texelSize.y ), dp ) +
	texture2DCompare( shadowMap, cubeToUV( bd3D + tx * offset.z, texelSize.y ), dp ) +
	texture2DCompare( shadowMap, cubeToUV( bd3D + ty * offset.z, texelSize.y ), dp ) +	
	texture2DCompare( shadowMap, cubeToUV( bd3D + tx * offset.x, texelSize.y ), dp ) +	
	texture2DCompare( shadowMap, cubeToUV( bd3D + ty * offset.x, texelSize.y ), dp )

) * ( 1.0 / 9.0 );
		
```

The result seems to look pretty good. Out of curiosity, what happened with the 4-sample approach?

MasterJames commented:
This is really awesome. I still have some belief that distance from camera should play a roll in detail optimization. Possible number of samples for instance?  Less samples are needed the further away from the camera for instance.

mrdoob commented:
> I looked over your 9-sample approach and it looks like you are not calculating the orthogonal vectors to be used as a new basis for the offset vectors (as suggested here). Was that intentional, or am I missing something?

It was just testing. I guessed sampling points in 3d space around the specific pixel should do the trick 😅

I'll try your code now.

> Out of curiosity, what happened with the 4-sample approach?

The 4 sample approach didn't result in good quality (not enough samples).

mrdoob commented:
@mkkellogg hmm, what's the `0.707107` for?

mkkellogg commented:
That is the length of the x & y components of a unit vector that has a z-value of zero. `tx`, `ty`, and `bd3D` create a new basis in which the offset vectors are created. Four of the offset vectors are unit vectors that are aligned/parallel with either the 'X' or 'Y' axis in this new basis:
```
( 1, 0, 0 )
( 0, 1, 0 )
( -1, 0, 0 )
( 0, -1, 0 )
```

The other four offset vectors are unit vectors in the XY plane in the new basis, but they each lie halfway in-between one of the above basis vectors and its "neighbor":

```
( 0.707107, 0.707107, 0 ) - Between positive X and positive Y axis
( -0.707107, 0.707107, 0 ) - Between negative X and positive Y axis
( -0.707107, -0.707107, 0 ) - Between negative X and negative Y axis
( 0.707107, -0.707107, 0 ) - Between positive X and negative Y axis
```

The second set is essentially the result of rotating the first set around the Z-axis in the new basis by 45 degrees.

The result is eight new offset vectors that can be appropriately multiplied by `tx` and `ty`, then scaled and added to `bd3D` to form the world-space offset vectors for sampling.

tschw commented:
@mrdoob 
what's the 0.707107 for?
```
0.707107 ~= sqrt( 1/2 ) = 1 / sqrt 2 = sin( 45° ) = cos( 45° )
```
which is the normalization factor for a vector with two one-valued components and others at zero or the vector components resulting from a polar angle of 45°.

mkkellogg commented:
@mrdoob 
Sorry if my previous comment was confusing (and long!). Thank you @tschw for being much more succinct :)

tschw commented:
@mkkellogg 
> Sorry if my previous comment was confusing (and long!).

Hey, I didn't mean to criticize! It's just one of those good-to-know constants when debugging computer graphics code, thus I added the math that produces it :-).

mrdoob commented:
Sounds good. Happy to replace the code. @tschw I assume the code look good to you?

tschw commented:
@mrdoob 
> @tschw I assume the code looks good to you?

Huh?! Which code? This commit: a445165f3023a8c5b34228025b51de8b3bf22eee ?

Quite frankly, I never studied that code too closely. Lame excuse: I don't really get it :blush:  

Well I think I *basically might do* . Four-tap thingy resembles an (approximately?) tetrahedral shape, I guess?

mrdoob commented:
No no, this other one by @mkkellogg: https://github.com/mrdoob/three.js/pull/7153#issuecomment-182486627

tschw commented:
@mrdoop
> No no, this other one

Ah! Oh yes, constructing an orthonormal basis seems clever. Makes more sense than to just sample into the blue overrating some of the samples due to spatial overlap.

@mkkellogg 
> The result seems to look pretty good. 

Ultimately, what looks good is good... 

> Out of curiosity, what happened with the 4-sample approach?

...The faster the better. Having those two orthogonal unit vectors in place, one could easily go like

```glsl
const float Scale = 1. / 9.;

const float AngleStep = 2. * PI / N_SECONDARY_SAMPLES;
const float AngleOffset = AngleStep * 0.5;

float result = texture2DCompare( shadowMap, cubeToUV( bd3D, texelSize.y ), dp ) * Scale;

for ( float i = 0; i < N_SECONDARY_SAMPLES; ++ i ) {

    float angle = AngleStep * i + AngleOffset;
    vec3 direction = bd3D + cos( angle ) * tx + sin( angle ) * ty; 
    result += texture2DCompare( shadowMap, cubeToUV( direction, texelSize.y ), dp ) * Scale;

}

return result;
```
to take N equidistant samples arranged in a circle plus one in the center (does that one even matter?) to see how low we can go before hardcoding something.

Note it's generally good practice to scale while accumulating, it's both numerically safer and a freebie on the GPU.

tschw commented:
@mkkellogg 
```glsl
// a is an arbitrary vector whose only purpose is to NOT be parallel with bd3D
vec3 a = bd3D + vec3( 1, 0, 0 ) + vec3( 0, 1, 0 );
```
On closer look, I'm afraid that this approach may cause visible artefacts in case `bd3D ~= (x,x,0)`.

Something like
```glsl
vec3 a = bd3D.yxz * vec3( 0.25, 0.5, 1. ) + vec3( 1, 0, 0 );
```
may be safer, but I'm not sure it's bulletproof.

@mrdoob 
Also, but that's a more general consideration: What's the unit of `shadowRadius`? Does it have defined semantics at all?

mrdoob commented:
> @mrdoob 
Also, but that's a more general consideration: What's the unit of shadowRadius? Does it have defined semantics at all?

Not yet... I just moved the poisson disk area out of the shader. It of course produces different results depending on the texture size.
